transform-plugins/src/main/java/co/cask/hydrator/plugin/ValueMapper.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.LookupConfig;
+import co.cask.cdap.etl.api.LookupProvider;
+import co.cask.cdap.etl.api.Transform;
+import co.cask.cdap.etl.api.TransformContext;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonSyntaxException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Transforms records using custom mapping provided by the config.
+ */
+@Plugin(type = "transform")
+@Name("ValueMapper")
+@Description("Maps and convert record values from input sources using a mapping dataset")
+public class ValueMapper extends Transform<StructuredRecord, StructuredRecord> {
+
+  private static final Gson GSON = new GsonBuilder().create();
+  private static final Logger LOG = LoggerFactory.getLogger(ValueMapper.class);
+  private final Config config;
+  private final Map<Schema, Schema> schemaCache = Maps.newHashMap();
+  private final Map<String, String> fieldsMapping = Maps.newHashMap();
+  private final Map<String, ValueMapperLookUp> datasetMapping =
+          Maps.newHashMap();
+  private JsonArray mappingArray;
+
+  // for unit tests, otherwise config is injected by plugin framework.
+  public ValueMapper(Config config) {
+    this.config = config;
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    LOG.debug("in transform start :" + input.get("name"));
+    Schema outputSchema = getOutputSchema(input.getSchema());
```

Comment:
Not sure here around how the input schema and output schema can be same. The value mapper will map the values of one or more input fields into output field either replacing the field in output schema or adding to it. So, when input schema is same as output schema -- this doesn' t adhere to the spec of value mapper then. Fix it!